### ISSUE TITLE
Detect and optimize away identity functions

### ISSUE THREAD

**Issue by RiscInside** [leanprover/lean4#1972]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Checked that your issue isn't already [filed](https://github.com/leanprover/lean4/issues).
    * Reduced the issue to a self-contained, reproducible test case.

The keyword I looked for is "identity", perhaps there is some other keyword I need to try?
Relevant Zulip thread: https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Code.20generation.20for.20what.20is.20essentially.20a.20no-op.20function

### Description

By identity functions, I am referring to the class of lean definitions that return one of their arguments unmodified. Unmodified here means that runtime representation of the argument is the same as that of returned value.

For instance, consider `Expr.leftExtend` function from this GitHub [gist](https://gist.github.com/RiscInside/6da4a485b83e6e7e55e7398b62005dce). It is defined as follows:

```lean
-- Left-extend `Expr n` to `Expr (n + 1)`
def Expr.leftExtend (e : Expr n) : Expr (n + 1) :=
  match e with
  | evar m m_lt_n   => evar m (Nat.lt_trans m_lt_n (Nat.lt_succ_self n))
  | emku _          => emku (n + 1)
  | euop u e        => euop u (e.leftExtend)
  | ebop b e₁ e₂    => ebop b (e₁.leftExtend) (e₂.leftExtend)
  | eabs e          => eabs (e.leftExtend)
  | elet e₁ e₂      => elet (e₁.leftExtend) (e₂.leftExtend)
  | ecas e e₁ e₂    => ecas (e.leftExtend) (e₁.leftExtend) (e₂.leftExtend)
  | efix e          => efix (e.leftExtend)
```
where `Expr` inductive type is defined as follows

```lean
-- Type family of expressions indexed by size of external context
inductive Expr : Nat → Type :=
  | evar (m : Nat) (_: m < n)                : Expr n -- de-brujin indexed variable
  | emku (n : Nat)                           : Expr n -- element of the unit type
  | euop (u : UnOp) (e : Expr n)             : Expr n -- unary operation
  | ebop (b : BinOp) (e₁ e₂ : Expr n)        : Expr n -- binary operation
  | eabs (e : Expr (n + 1))                  : Expr n -- λ-abstraction
  | elet (e₁ : Expr n) (e₂ : Expr (n + 1))   : Expr n -- let e₁ in e₂ (e₂ can refer to e₁ with evar)
  | ecas (e : Expr n) (e₁ e₂ : Expr (n + 1)) : Expr n -- case statement on sum type (encoded as two lambdas)
  | efix (e : Expr (n + 1))                  : Expr n -- fix combinator
  deriving DecidableEq
```

Here, runtime representation of `Expr n` should not depend on the value of `n`. As such, `Expr.leftExtend` function is implementable as an unsafe cast from `Expr n` to `Expr (n + 1)`. Ideally, compiler should be able to figure this on its own. As its stands however, compiler does indeed generate non-trivial code for `Expr.leftExtend`. This can be verified by asking Lean to dump C code in a file using `-c` command line option.

### Steps to Reproduce

1. Copy file from https://gist.github.com/RiscInside/6da4a485b83e6e7e55e7398b62005dce
2. Compile code with `-c` flag (e.g. `lean -c tiny.c Tiny.lean`)
3. Open the file and check that Lean compiler has indeed generated a rather non-trivial implementation of `Expr.leftExtend`

**Expected behavior:**

The lean compiler should be able to pick up on the fact that function is in fact an identity function - all calls to this function should be optimized away. Alternatively, the function generated should be a no-op (not a full recursive traversal).

**Actual behavior:**

Lean compiler generates a function that traverses the entire `Expr` tree.

**Reproduces how often:**

Identity functions like that are rare, but if there is one, lean compiler will generate code 100% of the time.

### Versions

You can get this information from copy and pasting the output of `lean --version`,
please include the OS and what version of the OS you're running.

```
notiurii@zen ~> lean --version
Lean (version 4.0.0, commit 7dbfaf9b7519, Release)
notiurii@zen ~> uname -a
Linux zen 5.15.0-56-generic #62~20.04.1-Ubuntu SMP Tue Nov 22 21:24:20 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

### Additional Information

I am not sure how this feature should work in general case. Perhaps Lean compiler should be able to identify inductive type parameters that do not influence runtime type representation. This can then be integrated in middle-end CSE or similar pass to identify which expressions will be common at runtime.

**Comment by gebner** [comment]
> Here, runtime representation of `Expr n` should not depend on the value of `n`.

You should know that it does though (but maybe we could change that).  The `n` is stored as a field in the expression:
```lean
set_option trace.compiler.ir.result true in
def f := @Expr.evar
```

If you want to guarantee an identical representation for expressions of different indices, then you should define them as subtypes.  Like `Expr n := { e : ExprRaw // e.varBound ≤ n }`.

**Comment by digama0** [comment]
> If you want to guarantee an identical representation for expressions of different indices, then you should define them as subtypes. Like `Expr n := { e : ExprRaw // e.varBound ≤ n }`.

Of course, if you use this representation, it also becomes trivial to get an O(1) conversion between `Expr n -> Expr m` since all the work is in the proof part. You would only need the identity function recognition trick for composite types built around these like `List (Expr n) -> List (Expr m)`.

**Comment by Kha** [comment]
Nothing to be done here unless we change how erasure works, in which case that would be tracked elsewhere
